File: 264.cpp
struct Solution {
    static vector<int> rec;
    static int idx[3];

    int nthUglyNumber(int n) {
        while (n >= this->rec.size())    {
            int l1 = this->rec[this->idx[0]] * 2,
                l2 = this->rec[this->idx[1]] * 3,
                l3 = this->rec[this->idx[2]] * 5;
            int x = min(l1, min(l2, l3));
            if (x == l1){
                ++this->idx[0];
            } else if (x == l2){
                ++this->idx[1];
            } else {
                ++this->idx[2];
            }
            if (x != this->rec.back()){
                this->rec.push_back(x);
            }
        }
        return this->rec[n - 1];
    }
};

vector<int> Solution::rec = {1};
int Solution::idx[3] = {0, 0, 0};

File: sol/264.py
class Solution:
    rec = [1]
    idx = [0, 0, 0]

    def nthUglyNumber(self, n: int) -> int:
        while n >= len(self.rec):
            l1 = self.rec[self.idx[0]] * 2
            l2 = self.rec[self.idx[1]] * 3
            l3 = self.rec[self.idx[2]] * 5
            x = min(l1, l2, l3)
            if x == l1:
                self.idx[0] += 1
            elif x == l2:
                self.idx[1] += 1
            else:
                self.idx[2] += 1
            if x != self.rec[-1]:
                self.rec.append(x)
        return self.rec[n - 1]
